Grbl_Esp32/src/Uart.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * UART driver that accesses the ESP32 hardware FIFOs directly.
+ */
+
+#include "Grbl.h"
+#include "Uart.h"
+
+#include "esp_system.h"
+#include "soc/uart_reg.h"
+#include "soc/io_mux_reg.h"
+#include "soc/gpio_sig_map.h"
+#include "soc/dport_reg.h"
+#include "soc/rtc.h"
+
+void Uart::flushTx() {
+    while (READ_PERI_REG(UART_STATUS_REG(_uart_num)) & (UART_TXFIFO_CNT_M | UART_ST_UTX_OUT)) {}
+}
+
+void Uart::setBaudRate(uint32_t baud_rate) {
+    uint32_t clk_div        = ((getApbFrequency() << 4) / baud_rate);
+    uint32_t clk_div_regval = ((clk_div >> 4) << UART_CLKDIV_S) | ((clk_div & UART_CLKDIV_FRAG_V) << UART_CLKDIV_FRAG_S);
+    WRITE_PERI_REG(UART_CLKDIV_REG(_uart_num), clk_div_regval);
+}
+
+Uart::Uart(int uart_num) : _uart_num(uart_num) {}
+
+void Uart::begin(unsigned long baudrate, uint32_t config, int8_t rxPin, int8_t txPin, bool invert, unsigned long timeout_ms) {
+    if (rxPin == -1 && txPin == -1) {
+        return;
+    }
+
+    switch (_uart_num) {
+        case 0:
+            DPORT_SET_PERI_REG_MASK(DPORT_PERIP_CLK_EN_REG, DPORT_UART_CLK_EN);
+            DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_RST_EN_REG, DPORT_UART_RST);
+            break;
+        case 1:
+            DPORT_SET_PERI_REG_MASK(DPORT_PERIP_CLK_EN_REG, DPORT_UART1_CLK_EN);
+            DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_RST_EN_REG, DPORT_UART1_RST);
+            break;
+        case 2:
+            DPORT_SET_PERI_REG_MASK(DPORT_PERIP_CLK_EN_REG, DPORT_UART2_CLK_EN);
+            DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_RST_EN_REG, DPORT_UART2_RST);
+            break;
+        default:
+            break;
+            // Can't happen
+    }
+
+    flushTx();
+    setBaudRate(baudrate);
+
+    // For now we hardcode the config as 8N1
+
+#define UART_8_DATABITS 3  // 0:5bits, 1:6bits, 2:7bits, 3:8bits
+#define UART_1_STOP_BIT 1  // 1:1bit, 2:1.5bits, 3:2bits
```

Comment:
Death to all defines... 
const int UART_8_DATABITS = 3;
const int UART_1_STOP_BIT = 1;

8N1 will also not work if we plan to use this for VFD's and other devices; it depends on the config there.

Reply:
I changed the code to handle arbitrary values.  Unfortunately that required the inclusion of some goofy code to work around some hardware bug that is mentioned vaguely in the ESP drivers, but not documented in the errata pdf.